language-agent/JVMMetric.proto

```diff
@@ -85,8 +86,22 @@ enum GCPhrase {
 }
 
 message Thread {
-  int64 liveCount = 1;
-  int64 daemonCount = 2;
-  int64 peakCount = 3;
+    int64 liveCount = 1;
+    int64 daemonCount = 2;
+    int64 peakCount = 3;
+    int64 deadlockedCount = 4;
+    int64 monitorDeadlockedCount = 5;
+    int64 newStateThreadCount = 6;
+    int64 runnableStateThreadCount = 7;
+    int64 blockedStateThreadCount = 8;
+    int64 waitingStateThreadCount = 9;
+    int64 timedWaitingStateThreadCount = 10;
+    int64 terminatedStateThreadCount = 11;
```

Comment:
In order to get this information, we need to get all active threads regularly and make statistics. I'm not sure our agent will accept this.

Reply:
My implemention.

```java
public enum ThreadProvider {
    INSTANCE;
    private final ThreadMXBean threadMXBean;

    ThreadProvider() {
        this.threadMXBean = ManagementFactory.getThreadMXBean();
    }

    public Thread getThreadMetrics() {
        int newThreadCount = 0;
        int runnableThreadCount = 0;
        int blockedThreadCount = 0;
        int waitThreadCount = 0;
        int timeWaitThreadCount = 0;
        int terminatedThreadCount = 0;

        ThreadInfo[] threadInfos = threadMXBean.getThreadInfo(threadMXBean.getAllThreadIds());
        if (threadInfos != null) {
            for (ThreadInfo threadInfo : threadInfos) {
                if (threadInfo != null) {
                    switch (threadInfo.getThreadState()) {
                        case NEW:
                            newThreadCount++;
                            break;
                        case RUNNABLE:
                            runnableThreadCount++;
                            break;
                        case BLOCKED:
                            blockedThreadCount++;
                            break;
                        case WAITING:
                            waitThreadCount++;
                            break;
                        case TIMED_WAITING:
                            timeWaitThreadCount++;
                            break;
                        case TERMINATED:
                            terminatedThreadCount++;
                            break;
                        default:
                            break;
                    }
                } else {
                    terminatedThreadCount++;
                }
            }
        }

        int threadCount = threadMXBean.getThreadCount();
        int daemonThreadCount = threadMXBean.getDaemonThreadCount();
        int peakThreadCount = threadMXBean.getPeakThreadCount();
        int deadlocked = threadMXBean.findDeadlockedThreads() != null ? threadMXBean.findDeadlockedThreads().length : 0;
        int monitorDeadlocked = threadMXBean.findMonitorDeadlockedThreads() != null ?
                threadMXBean.findMonitorDeadlockedThreads().length : 0;
        return Thread.newBuilder().setLiveCount(threadCount)
                .setDaemonCount(daemonThreadCount)
                .setPeakCount(peakThreadCount)
                .setDeadlocked(deadlocked)
                .setMonitorDeadlocked(monitorDeadlocked)
                .setNewThreadCount(newThreadCount)
                .setRunnableThreadCount(runnableThreadCount)
                .setBlockedThreadCount(blockedThreadCount)
                .setWaitThreadCount(waitThreadCount)
                .setTimeWaitThreadCount(timeWaitThreadCount)
                .setTerminatedThreadCount(terminatedThreadCount)
                .build();
    }

}
```

Reply:
![image](https://user-images.githubusercontent.com/11993575/124253610-cb57ca80-db5a-11eb-86f5-e8f3349a559d.png)

Reply:
We may need a benchmark about this way, could you use `jmh` to run this? I want to see whether there is a performance impact.

Reply:
Please don't resolve it. There is an comment.

Reply:
> We may need a benchmark about this way, could you use `jmh` to run this? I want to see whether there is a performance impact.

OK, I will test it later.

Reply:
The `prometheus/client_java` use ThreadMXBean and ClassLoadingMXBean to get metric.


- [ThreadExports.java](https://github.com/prometheus/client_java/blob/master/simpleclient_hotspot/src/main/java/io/prometheus/client/hotspot/ThreadExports.java)
- [ClassLoadingExports.java](https://github.com/prometheus/client_java/blob/master/simpleclient_hotspot/src/main/java/io/prometheus/client/hotspot/ClassLoadingExports.java)

Reply:
I know what you mean. In theory, the performance loss of your implementation is relatively low.

![image](https://user-images.githubusercontent.com/28091237/124270680-1c71b980-db6f-11eb-9495-cb13a35bc855.png)

Reply:
ThreadProviderBenchmark

```java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

package org.apache.skywalking.apm.agent.core.jvm.thread;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.profile.GCProfiler;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;

import java.util.concurrent.TimeUnit;

public class ThreadProviderBenchmark {

    @Benchmark
    @Fork(value = 1, warmups = 1)
    @OutputTimeUnit(TimeUnit.SECONDS)
    @BenchmarkMode(Mode.Throughput)
    public void getThreadMetrics() {
        ThreadProvider.INSTANCE.getThreadMetrics();
    }

    public static void main(String[] args) throws Exception {
        Options opt = new OptionsBuilder().include(ThreadProviderBenchmark.class.getSimpleName())
                .addProfiler(GCProfiler.class)
                .build();
        new Runner(opt).run();
    }

    /**
     * # JMH version: 1.21
     * # VM version: JDK 1.8.0_231, Java HotSpot(TM) 64-Bit Server VM, 25.231-b11
     * # VM invoker: /Library/Java/JavaVirtualMachines/jdk1.8.0_231.jdk/Contents/Home/jre/bin/java
     * # VM options: -javaagent:/Users/switch/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/211.7442.40/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=65029:/Users/switch/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/211.7442.40/IntelliJ IDEA.app/Contents/bin -Dfile.encoding=UTF-8
     * # Warmup: 5 iterations, 10 s each
     * # Measurement: 5 iterations, 10 s each
     * # Timeout: 10 min per iteration
     * # Threads: 1 thread, will synchronize iterations
     * # Benchmark mode: Throughput, ops/time
     *
     * Benchmark                                  Mode  Cnt      Score      Error  Units
     * ThreadProviderBenchmark.getThreadMetrics  thrpt    5  12275.329 ± 5132.424  ops/s
     */
}
```

Reply:
> ThreadProviderBenchmark

I will compare with `prometheus/client_java` ThreadExports.java .

Reply:
## ClassProvider

```java
package org.apache.skywalking.apm.agent.core.jvm.clazz;

import java.lang.management.ClassLoadingMXBean;
import java.lang.management.ManagementFactory;
import org.apache.skywalking.apm.network.language.agent.v3.Class;

public enum ClassProvider {
    INSTANCE;
    private final ClassLoadingMXBean classLoadingMXBean;

    ClassProvider() {
        this.classLoadingMXBean = ManagementFactory.getClassLoadingMXBean();
    }

    public Class getClassMetrics() {
        int loadedClassCount = classLoadingMXBean.getLoadedClassCount();
        long unloadedClassCount = classLoadingMXBean.getUnloadedClassCount();
        long totalLoadedClassCount = classLoadingMXBean.getTotalLoadedClassCount();
        return Class.newBuilder().setLoadedClassCount(loadedClassCount)
                .setUnloadedClassCount(unloadedClassCount)
                .setTotalLoadedClassCount(totalLoadedClassCount)
                .build();
    }

}
```

## ClassProviderBenchmark

```java
package org.apache.skywalking.apm.agent.core.jvm.clazz;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;

import java.util.concurrent.TimeUnit;

public class ClassProviderBenchmark {

    @Benchmark
    @Fork(value = 1, warmups = 1)
    @OutputTimeUnit(TimeUnit.SECONDS)
    @BenchmarkMode(Mode.Throughput)
    public void getThreadMetrics() {
        ClassProvider.INSTANCE.getClassMetrics();
    }

    public static void main(String[] args) throws Exception {
        Options opt = new OptionsBuilder().include(ClassProviderBenchmark.class.getSimpleName())
                .build();
        new Runner(opt).run();
    }

    /**
     * # JMH version: 1.21
     * # VM version: JDK 1.8.0_231, Java HotSpot(TM) 64-Bit Server VM, 25.231-b11
     * # VM invoker: /Library/Java/JavaVirtualMachines/jdk1.8.0_231.jdk/Contents/Home/jre/bin/java
     * # VM options: -javaagent:/Users/switch/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/211.7442.40/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=65164:/Users/switch/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/211.7442.40/IntelliJ IDEA.app/Contents/bin -Dfile.encoding=UTF-8
     * # Warmup: 5 iterations, 10 s each
     * # Measurement: 5 iterations, 10 s each
     * # Timeout: 10 min per iteration
     * # Threads: 1 thread, will synchronize iterations
     * # Benchmark mode: Throughput, ops/time
     *
     * Benchmark                                 Mode  Cnt        Score         Error  Units
     * ClassProviderBenchmark.getThreadMetrics  thrpt    5  6461016.100 ± 1453485.840  ops/s
     */
}
```

Reply:
There are some issues in the benchmark:

- the fork `value=1` and `warmups=1` are too small IMO.
- The result of the tested method `ClassProvider.INSTANCE.getClassMetrics()` and `ThreadProvider.INSTANCE.getThreadMetrics()` are not consumed, which might be optimized into NOOP by JVM.

If this is a critical path and the benchmark result really matters, we should redo with 👆 fixed, otherwise this PR also looks good to me.

Reply:
> There are some issues in the benchmark:
> 
> * the fork `value=1` and `warmups=1` are too small IMO.
> * The result of the tested method `ClassProvider.INSTANCE.getClassMetrics()` and `ThreadProvider.INSTANCE.getThreadMetrics()` are not consumed, which might be optimized into NOOP by JVM.
> 
> If this is a critical path and the benchmark result really matters, we should redo with 👆 fixed, otherwise this PR also looks good to me.

ok, I will test again.

Reply:
> There are some issues in the benchmark:
> 
> * the fork `value=1` and `warmups=1` are too small IMO.
> * The result of the tested method `ClassProvider.INSTANCE.getClassMetrics()` and `ThreadProvider.INSTANCE.getThreadMetrics()` are not consumed, which might be optimized into NOOP by JVM.
> 
> If this is a critical path and the benchmark result really matters, we should redo with 👆 fixed, otherwise this PR also looks good to me.

Increase fork `value` to 5 and `warmups` to 3.
Use `Blackhole` consume the result.

@kezhenxu94 please review again.

## ThreadProviderBenchmark

```java
package org.apache.skywalking.apm.agent.core.jvm.thread;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.infra.Blackhole;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;

import java.util.concurrent.TimeUnit;

public class ThreadProviderBenchmark {

    @Benchmark
    @Fork(value = 5, warmups = 3)
    @OutputTimeUnit(TimeUnit.SECONDS)
    @BenchmarkMode(Mode.Throughput)
    public void getThreadMetrics(Blackhole bh) {
        bh.consume(ThreadProvider.INSTANCE.getThreadMetrics());
    }

    public static void main(String[] args) throws Exception {
        Options opt = new OptionsBuilder().include(ThreadProviderBenchmark.class.getSimpleName())
                .build();
        new Runner(opt).run();
    }

    /**
     * # JMH version: 1.21
     * # VM version: JDK 1.8.0_231, Java HotSpot(TM) 64-Bit Server VM, 25.231-b11
     * # VM invoker: /Library/Java/JavaVirtualMachines/jdk1.8.0_231.jdk/Contents/Home/jre/bin/java
     * # VM options: -javaagent:/Users/switch/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/211.7442.40/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=52623:/Users/switch/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/211.7442.40/IntelliJ IDEA.app/Contents/bin -Dfile.encoding=UTF-8
     * # Warmup: 5 iterations, 10 s each
     * # Measurement: 5 iterations, 10 s each
     * # Timeout: 10 min per iteration
     * # Threads: 1 thread, will synchronize iterations
     * # Benchmark mode: Throughput, ops/time
     * # Benchmark: org.apache.skywalking.apm.agent.core.jvm.thread.ThreadProviderBenchmark.getThreadMetrics
     *
     * Benchmark                                  Mode  Cnt       Score      Error  Units
     * ThreadProviderBenchmark.getThreadMetrics  thrpt   25  247393.607 ± 2493.640  ops/s
     */
}
```

## ClassProviderBenchmark
```java
package org.apache.skywalking.apm.agent.core.jvm.clazz;

import org.openjdk.jmh.annotations.Benchmark;
import org.openjdk.jmh.annotations.BenchmarkMode;
import org.openjdk.jmh.annotations.Fork;
import org.openjdk.jmh.annotations.Mode;
import org.openjdk.jmh.annotations.OutputTimeUnit;
import org.openjdk.jmh.infra.Blackhole;
import org.openjdk.jmh.runner.Runner;
import org.openjdk.jmh.runner.options.Options;
import org.openjdk.jmh.runner.options.OptionsBuilder;

import java.util.concurrent.TimeUnit;

public class ClassProviderBenchmark {

    @Benchmark
    @Fork(value = 5, warmups = 3)
    @OutputTimeUnit(TimeUnit.SECONDS)
    @BenchmarkMode(Mode.Throughput)
    public void getThreadMetrics(Blackhole bh) {
        bh.consume(ClassProvider.INSTANCE.getClassMetrics());
    }

    public static void main(String[] args) throws Exception {
        Options opt = new OptionsBuilder().include(ClassProviderBenchmark.class.getSimpleName())
                .build();
        new Runner(opt).run();
    }

    /**
     # JMH version: 1.21
     # VM version: JDK 1.8.0_231, Java HotSpot(TM) 64-Bit Server VM, 25.231-b11
     # VM invoker: /Library/Java/JavaVirtualMachines/jdk1.8.0_231.jdk/Contents/Home/jre/bin/java
     # VM options: -javaagent:/Users/switch/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/211.7442.40/IntelliJ IDEA.app/Contents/lib/idea_rt.jar=52931:/Users/switch/Library/Application Support/JetBrains/Toolbox/apps/IDEA-U/ch-0/211.7442.40/IntelliJ IDEA.app/Contents/bin -Dfile.encoding=UTF-8
     # Warmup: 5 iterations, 10 s each
     # Measurement: 5 iterations, 10 s each
     # Timeout: 10 min per iteration
     # Threads: 1 thread, will synchronize iterations
     # Benchmark mode: Throughput, ops/time
     *
     * Benchmark                                 Mode  Cnt        Score      Error  Units
     * ClassProviderBenchmark.getThreadMetrics  thrpt   25  6542809.978 ± 8794.520  ops/s
     */
}
```